## TP06 : lecture-écriture dans un fichier  - tableaux à deux ou trois dimensions   
**codage d'une image - traitement d'images**

- Éléments de cours sur lecture et écriture dans un fichier de texte :  <à venir>  
- Éléments de cours sur les chaînes de caractères :  https://capytale2.ac-paris.fr/web/c-auth/list?returnto=/web/code/1db0-160848
        
    

Raccourcis pour Jupyter :
<ul>
    <li>Maj + Enter : exécute la cellule courante ;</li>
    <li>Esc b : crée une cellule de code au-dessous de la cellule courante ('b' pour 'below') ;</li>
    <li>Alt + Enter : exécute la cellule courante et créé une cellule de code à sa suite ;</li>
    <li>Esc + m : transforme une cellule de code (mode 'Code') en cellule de texte (mode 'Markdown') ;</li>
    <li>Esc + y : transforme cellule de texte (mode 'Markdown') en une cellule de code (mode 'Code').</li>
</ul>

## Note préliminaire : fichiers nécessaires pour ce TP

In [1]:
liste_fichiers_TP06 = ['haiku.txt',
                       'LettreB.txt',
                       'tp06-fruits.txt',
                       'alphabet_pixels.txt',
                       'pomme.pbm',
                       'lettreG.pbm',
                       'Flag_of_Sweden320x200.png']

**Si vous travaillez sur machine, avec IDLE ou Spyder, ces fichiers doivent être présents dans le <u>même répertoire que le fichier .py qui les utilise</u>** :   
- pour une utilisation sur un PC du lycée : copier le sous-dossier `TP06` situé dans le répertoire `_travail` de la classe (dossier réservé à l'informatique de tronc commun) et collez-le dans votre répertoire personnel, dans le sous-dossier `Info` que vous utilisez habituellement.
- pour une utilisation sur votre PC personnel : téléchargez les fichiers du TP06 depuis Moodle et copiez-les dans le même répertoire que le programme que vous écrirez et qui les utilise.

<span style="background-color:yellow;color:red">**Si vous travaillez en ligne sur Capytale, <u>exécutez la cellule suivante pour vérifier que les fichiers utilisés sont bien accessibles</u>**</span> :  

In [2]:
import os
liste_fichiers_TP06 = ['haiku.txt', \
                       'LettreB.txt', \
                       'tp06-fruits.txt', \
                       'alphabet_pixels.txt', \
                       'pomme.pbm', \
                       'lettreG.pbm', \
                       'Flag_of_Sweden320x200.png']
d = {'haiku.txt': 'haiku.txt', 'lettreb.txt': 'LettreB.txt', 'monfichier.txt': 'monfichier.txt', \
     'tp06-fruits.txt': 'tp06-fruits.txt', 'alphabet_pixels.txt': 'alphabet_pixels.txt', \
     'pomme.pbm_.txt': 'pomme.pbm', 'lettreG.pbm_.txt': 'lettreG.pbm', \
     'flag_of_Sweden320x200.png': 'Flag_of_Sweden320x200.png'}
for fichier, nomfichier in d.items():
    if fichier in os.listdir('.'):
        os.rename(fichier, nomfichier)
for f in liste_fichiers_TP06:
    if f not in os.listdir('.'): print("!!!!! fichier", f, "absent.")
dispo = len([f for f in liste_fichiers_TP06 if f in os.listdir('.')])
print(f'{dispo} fichier(s) disponibles sur {len(liste_fichiers_TP06)} fichier(s) requis.')

7 fichier(s) disponibles sur 7 fichier(s) requis.


In [3]:
os.listdir()

['haiku.txt', 'LettreB.txt', 'tp06-fruits.txt', 'alphabet_pixels.txt', 'pomme.pbm', 'lettreG.pbm', 'Flag_of_Sweden320x200.png', 'monfichier.txt', 'LettreL.txt', 'myfile.ext', 'malettreA.pbm']

## Introduction

De façon générale, en informatique, une image sous forme brute, peut être représentée sous la forme d'une matrice de pixels, l'information codée dans chaque pixel étant la couleur de ce pixel :  
* s'il s'agit d'une image en noir et blanc, chaque pixel pourra prendre deux valeurs, l'une codant la couleur &laquo;blanc&raquo;, l'autre la couleur &laquo;noir&raquo; ;  
* s'il s'agit d'une image en **niveaux de gris**, chaque pixel pourra prendre une valeur dans un intervalle de valeurs, entières ou flottantes, les deux valeurs extrêmes codant le noir et le blanc, les valeurs intermédiaires codant les différents niveaux de gris ;  
* s'il s'agit d'une image en couleur, il faut définir un système de représentation des couleurs, ici la réprésentation sous la forme de trois composantes, rouge, vert et bleu, qui permet, par synthèse additive, de produire les différentes couleurs, sur un écran, en allumant avec des intensités différentes trois leds contingües, rouge, verte et bleue. Ces trois intensités seront codées par un triplet d'entiers ou de flottants prenant leurs valeurs dans un intervalle (dans ce TP il s'agirait d'un intervalle de $256$ valeurs entières, $[\![0;255]\!]$, codable sur $8$ bits).

Ci-dessous, une image rudimentaire de &laquo;pomme&raquo;, en noir et blanc, sous la forme d'une matrice $16\times 16$ de pixels.

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_pomme_NB.png" width="100"/>

On s'intéresse dans ce TP à l'encodage d'une image sous la forme d'une matrice de pixels, d'une part pour lire ou sauvegarder une image sur un support physique, d'autre part pour effectuer des traitements sur une image.

## Exercice 1 encodage d'une police de caractère sous la forme d'un fichier texte

Lorsque l'on s'intéresse à la l'affichage d'un texte sur un écran, on définit une ***police*** de caractères, dans laquelle on convient de l'image qui sera affichée pour représenter chaque caractère du texte.

Voici une police de caractères rudimentaire pour les lettres majuscules de l'alphabet latin :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_alphabet_complet-NB.png" width = "90%"/><br/>
- source : https://www.millenium.org/news/125335.html

Dans ce modèle, chaque lettre sera représentée par une matrice de pixels de taille $7\times 5$, à $7$ lignes et $5$ colonnes.

Si on représente les pixels noirs par la lettre `'x'` et les pixels blancs par des espaces `' '`, on peut représenter comme suit la lettre majuscule &laquo; A &raquo; :

In [4]:
A = "     \n xxx \nx   x\nx   x\nxxxxx\nx   x\nx   x\nx   x\n"

On a utilisé le caractère de passage à la ligne `'\n'` afin que l'on puisse obtenir l'affichage de la lettre &laquo; A &raquo; avec l'instruction suivante :

In [5]:
print(A)

     
 xxx 
x   x
x   x
xxxxx
x   x
x   x
x   x



**Q1** En utilisant uniquement les opérations `*` et `+` définies pour les chaînes de caractères, définir de même une chaîne de caractères, nommée `L`, représentant la lettre &laquo; L &raquo; dans ce modèle, puis l'afficher.

In [6]:
L = ... ## <- à compléter
print(L)

Ellipsis


***Q1 corrigé***

Dans le modèle, toute lettre est représentée par sept lignes de cinq caractères, égaux à `'x'` ou `' '`.

In [7]:
L = 6 * ('x' + 4 * ' ' + '\n') + 5 * 'x' ## <- à compléter
print(L)

x    
x    
x    
x    
x    
x    
xxxxx


La chaîne `L` construite, sous sa forme brute, est la suivante :

In [8]:
L

'x    \nx    \nx    \nx    \nx    \nx    \nxxxxx'

**[préliminaire pour la question Q2]** On rappelle ci-dessous la procédure la plus élémentaire pour ouvrir un fichier texte à l'aide Python.  
On ouvre dans l'exemple, un fichier contenant un <a href="https://fr.wikipedia.org/wiki/Ha%C3%AFku">haïku célèbre du maître Bashō</a>.<br/>
Ici le chemin vers le fichier comprend uniquement le nom du fichier suivi de son extension, car le fichier a été chargé dans le même répertoire que ce notebook.

In [9]:
f = open('haiku.txt')
texte_haiku = f.read()
f.close()

Le contenu du fichier a été chargé sous la forme d'une unique chaîne de caractères, `texte_haiku`, que l'on peut visualiser sous sa forme brute :

In [10]:
texte_haiku

"Un vieil étang\nUne grenouille qui plonge,\nLe bruit de l'eau."

ou afficher à l'aide de la fonction `print` :

In [11]:
print(texte_haiku) ## la fonction print interpète les caractères spéciaux, ici '\n' qui encode un passage à la ligne.

Un vieil étang
Une grenouille qui plonge,
Le bruit de l'eau.


*Note :* si le fichier est ouvert sur une machine exécutant Windows, il faudra rajouter la spécificatiion de l'encodage du fichier, et utiliser l'instruction  `open('haiku.txt', encoding = 'utf-8')` au lieu de `open('haiku.txt')`, afin que les lettres accentuées soient correctement décodées.

**Q2** Un fichier texte  <code>LettreB.txt</code> a été téléchargé dans le répertoire courant de ce notebook.  
Lire le contenu de ce fichier, et le stocker dans une variable `B`.  
Vérifier le contenu brut de cette variable, puis afficher la lettre B.

* ouverture et lecture du contenu du fichier

* Vérification du contenu brut de la variable `B`

* Affichage de la chaîne avec la fonction `print`

***Q2 corrigé***

* ouverture et lecture du contenu du fichier

In [12]:
f = open('LettreB.txt')
B = f.read()
f.close()

* Vérification du contenu brut de la variable `B`

In [13]:
B

'xxxx \nx   x\nx   x\nxxxx \nx   x\nx   x\nxxxx '

* Affichage de la chaîne avec la fonction `print`

In [14]:
print(B)

xxxx 
x   x
x   x
xxxx 
x   x
x   x
xxxx 


**[préliminaire pour la question Q3]** On rappelle ci-dessous la procédure pour créer et écrire dans un fichier texte à l'aide Python.

In [15]:
f = open('monfichier.txt', 'w')
f.write('Hello World !')
f.write('\nAuteur·rice : \tun·e apprenti·e informaticien·ne.')
f.close()

On a là créé un fichier dont le contenu est (visualisé avec un éditeur de texte basique, comme "Bloc notes" sur Windows) :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_edition_monfichier-txt_v3.png"/>

Il a fallu écrire dans le fichier tous les caractères spéciaux de passage à la ligne, `\n`, et de tabulation, `\t`, pour obtenir ce résultat.

**Q3** Créer dans le répertoire courant de ce notebook, un fichier texte, <code>LettreL.txt</code>, qui s'afficherait, dans un éditeur de texte, avec la forme suivante

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_edition_LettreL-txt.png"/>

* Vérifier que le fichier a bien été créé et est conforme aux attendus, en décommentant les trois cellules suivantes pour ensuite les exécuter 

In [16]:
## décommenter puis exécuter les trois lignes suivantes : on ouvrira ainsi le fichier créé pour récupérer son contenu sous
## la forme d'une chaîne de caractères nommée contenu_fichier :

#f = open('LettreL.txt')
#contenu_fichier = f.read()
#f.close()

In [17]:
## décommenter et exécuter la ligne suivante pour examiner le contenu brut du fichier

#contenu_fichier

In [18]:
## décommenter et exécuter la ligne suivante pour afficher le contenu du fichier à l'aide de la fonction print

#print(contenu_fichier)

***Q3 corrigé***

In [19]:
f = open('LettreL.txt', mode = 'w')
f.write(L) ## <-- on utilise la variable L (chaîne de caractères) définie à la question 1
f.close()

* Vérifier que le fichier a bien été créé et est conforme aux attendus, en décommentant les trois cellules suivantes pour ensuite les exécuter 

In [20]:
## décommenter puis exécuter les trois lignes suivantes : on ouvrira ainsi le fichier créé pour récupérer son contenu sous
## la forme d'une chaîne de caractères :

f = open('LettreL.txt')
contenu_fichier = f.read()
f.close()

In [21]:
## décommenter et exécuter la ligne suivante pour examiner le contenu brut du fichier

contenu_fichier

'x    \nx    \nx    \nx    \nx    \nx    \nxxxxx'

In [22]:
## décommenter et exécuter la ligne suivante pour afficher le contenu du fichier à l'aide de la fonction print

print(contenu_fichier)

x    
x    
x    
x    
x    
x    
xxxxx


**[préliminaire pour la question Q4]** On rappelle la façon d'ouvrir un fichier texte de sorte pour en charger le contenu sous la forme d'une liste dont chaque élément est une chaîne de caractères correspondant à une ligne de ce fichier. <br/>
Sur l'exemple, on charge le contenu d'un fichier `tp06-fruits.txt` dont chaque ligne contient le nom d'un fruit.

In [23]:
f = open('tp06-fruits.txt')
liste_lignes = f.readlines()
f.close()

In [24]:
liste_lignes

['Abricot\n', 'Cerise\n', 'Citron\n', 'Fraise\n', 'Framboise\n', 'Kiwi\n', 'Myrtille\n', 'Poire']

In [25]:
print(liste_lignes)

['Abricot\n', 'Cerise\n', 'Citron\n', 'Fraise\n', 'Framboise\n', 'Kiwi\n', 'Myrtille\n', 'Poire']


**[préliminaire pour la question Q4]** On rappelle (*cf.* **Annexe A** du programme d'informatique de tronc commun première et deuxième année : <a href="https://github.com/lbaille20/Info1A_2021-22/raw/main/Annexe_A_Langage_Python.pdf">ici</a>) que l'on peut copier une portion (***tranche***) d'une chaîne de caractères `ch` à l'aide du *slicing* :

In [26]:
ch = 'abracadabra'

Avec la syntaxe `ch[a:b]`, où $0\leqslant a \leqslant b <\texttt{len(ch)}$, on obtient une copie des caractères dont la position $i$ est telle que $a\leqslant i< b$.  
Il est **important de se rappeler** que, dans ces conditions, **la tranche obtenue est de longueur $b-a$**.

In [27]:
portion = ch[1:7] ## copie des caractères de la chaîne ch, dont la position i est telle que 1 <= i < 7

In [28]:
portion # de longueur 7 - 1 = 6 (la tranche comportera six caractères)

'bracad'

**Q4** Utiliser le *slicing* pour construire, à partir de la liste `liste_lignes`, une liste `liste_fruits` dans laquelle on aura supprimé les caractères de passage à la ligne.  
Ainsi, la liste `liste_fruits` aura l'aspect suivant : `['Abricot', 'Cerise', ..., 'Myrtille', 'Poire']`.  
On proposera une solution dans laquelle on accumulera, avec la méthode `.append` les noms de fruits dans la liste `liste_fruits`, initialement vide, et une solution utilisant une compréhension de liste et l'opération de concaténation `'+'` pour les listes.  
Dans la mise en œuvre de la deuxième solution on donne, pour obtenir une copie d'une chaîne de caractères privée de son dernier caractère `ch`, la construction `ch[:-1]`, qui est un abrégé de la construction `ch[0:len(ch)-1]`, et l'abrégé `L[-1]`, en remplacement de `L[len(L)-1]`, pour obtenir le dernier élément d'une liste `L`.

***Q4 corrigé***

* solution 1

In [29]:
n = len(liste_lignes)

liste_fruits = []
for k in range(n - 1):         ## parcours de la liste liste_lignes privée de son dernier élément
    nom_fruit = liste_lignes[k]      ## le nom du fruit lu se termine par un caractère de fin de ligne
    lg = len(nom_fruit)              ## longueur de la chaîne nom_fruit, caractère de fin de ligne inclus
    liste_fruits.append(nom_fruit[0:lg - 1]) ## ajout du nom de fruit en supprimant le dernier caractère '\n'
liste_fruits.append(liste_lignes[n - 1])## on ajoute le nom du dernier fruit, qui n'est pas suivi d'un '\n'

In [30]:
# Vérification
liste_fruits

['Abricot', 'Cerise', 'Citron', 'Fraise', 'Framboise', 'Kiwi', 'Myrtille', 'Poire']

* solution 2

In [31]:
liste_fruits = [liste_lignes[k][:-1] for k in range(len(liste_lignes) - 1)] + [ liste_lignes[-1] ]
print(liste_fruits)

['Abricot', 'Cerise', 'Citron', 'Fraise', 'Framboise', 'Kiwi', 'Myrtille', 'Poire']


***\[ne pas faire ce qui suit - fin de l'exercice 1 - pendant le temps du TP: <span style="background-color:yellow;color:red"> passer à l'exercice 2</span>\]***

**Q5**  On a défini dans un unique fichier texte, `alphabet_pixels.txt`, les $26$ lettres de l'alphabet encodés comme précédemment.  


<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_edition_alphabet_pixel-txt.png"/>

**Q5.a** Sans compter les caractères de passage à la ligne, combien de caractères composent chaque ligne du fichier `alphabet_pixels.txt`, en rappelant que chaque lettre est codée par une matrice de $7\times 5$ caractères et en remarquant que chaque lettre est séparée de sa suivante par une espace ?

***Q5.a corrigé***

Il y a 26 lettres codées et une espace entre chaque. Chaque ligne du codage d'une lettre comporte 5 caractères.  
Donc chaque ligne du fichier est composée de $26 \times 5 + 25 = 155$ caractères (caractère de passage à la ligne, `'\n'` exclu).  

**Q5.b** Ouvrir le fichier `alphabet_pixels.txt` (situé dans même répertoire que ce notebook) et en récupérer le contenu sous la forme de la liste de ses lignes dans une variable `L`.

***Q5.b corrigé***

In [32]:
f = open('alphabet_pixels.txt')
L = f.readlines()
f.close()

*Note :* on peut vérifier que chaque ligne lue comporte 156 caractères (caractère de passage à la ligne inclus), sauf la dernière, qui ne comporte pas de caractère de passage à la ligne.

In [33]:
L[0]

' xxx  xxxx   xxx  xxxx  xxxxx xxxxx  xxx  x   x  xxx    xxx x   x x     x   x x   x  xxx  xxxx   xxx  xxxx   xxxx xxxxx x   x x   x x   x x   x x   x xxxxx\n'

In [34]:
len(L[0])

156

In [35]:
for ligne in L:
    print(len(ligne))

156
156
156
156
156
156
155


**Q5.c** Quelle tranche de la première ligne permet de récupérer la première ligne du codage de la lettre &laquo; A &raquo; ? 

***Q5.c corrigé***

Il s'agit de copier les 5 premiers caractères de la première ligne :

In [36]:
L[0][0:5]

' xxx '

**Q5.d** Récupérer depuis la liste `L`,  sous la forme d'une liste nommée `listeA`, les sept lignes du codage de la lettre &laquo; A &raquo;.  
On proposera deux solutions, avec et sans compréhension de liste.

***Q5.d corrigé***

Il faut récupérer dans une liste, `listeA`, les cinq premiers caractères de chaque ligne.

* solution 1

In [37]:
listeA = []
for ligne in L:
    listeA.append(ligne[0:5])

In [38]:
## vérification
listeA

[' xxx ', 'x   x', 'x   x', 'xxxxx', 'x   x', 'x   x', 'x   x']

* solution 2

In [39]:
listeA = [ligne[0:5] for ligne in L]

In [40]:
## vérification
listeA

[' xxx ', 'x   x', 'x   x', 'xxxxx', 'x   x', 'x   x', 'x   x']

**Q5.e** Concaténer les sept éléments de la liste `listeA` en une chaîne `lettreA`, permettant d'afficher la lettre &laquo; A &raquo; comme en introduction, à l'aide de l'instruction `print(lettreA)`. 

***Q5.e corrigé***

In [41]:
lettreA = ""
for ligne in listeA:
    lettreA = lettreA + ligne + '\n'

In [42]:
## affichage de la lettre A
print(lettreA)

 xxx 
x   x
x   x
xxxxx
x   x
x   x
x   x



**Q5.f** Reprendre la question **5.d** pour obtenir le codage des lettres &laquo; B &raquo; et &laquo; C &raquo;, sous la forme de deux listes de chaînes de caractères, `listeB` et `listeC`.   
On utilisera des compréhension de listes.

***Q5.f corrigé***

Sur chaque ligne, les six premiers caractères sont dédiés au codage de la première ligne de la lettre A et à l'espace séparant la lettre A de la lettre B.  
Le codage de la première ligne de la lettre B commence donc au caractère d'indice 6 dans la chaîne L[0], et comporte cinq caractères.    
Il en est de même sur les six autres lignes, `L[1]`, ..., `L[6]`.

In [43]:
listeB = [L[k][6:6 + 5] for k in range(len(L))]
## ou bien
listeB = [ligne[6:6 + 5] for ligne in L]

In [44]:
## Vérification
print(listeB)

['xxxx ', 'x   x', 'x   x', 'xxxx ', 'x   x', 'x   x', 'xxxx ']


De façon générale, le codage des $n$ premières lettres de l'alphabet, espace de séparation entre lettres inclus, correspond aux $6n$ premiers caractères de chaque ligne, les cinq caractères suivants correspondant au codage de la $(n+1)^{\rm ème}$ lettre de l'alphabet.  
Ainsi, pour la lettre C, qui est la troisième lettre de l'alphabet :

In [45]:
listeC = [ligne[6 * 2:6 * 2 + 5] for ligne in L]

In [46]:
## Vérification
print(listeC)

[' xxx ', 'x   x', 'x    ', 'x    ', 'x    ', 'x   x', ' xxx ']


**Q5.g** Afin de composer un texte avec toutes les lettres de l'alphabet, on souhaite définir un dictionnaire dont les clés sont données par la liste suivante :

In [47]:
majuscules = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', \
              'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']

*Note :* cette liste peut être obtenue par les instructions suivantes

In [48]:
from string import ascii_uppercase as Majuscules

In [49]:
majuscules = list(Majuscules)

**Q5.g.i** Définir un dictionnaire `Dlettres` dont les clés sont les éléments de la liste `majuscules` et les valeurs sont des listes vides.   

***Q5.g.i corrigé***

In [50]:
Dlettres = {}
for cle in majuscules:
    Dlettres[cle] = []

In [51]:
## Vérification
Dlettres

{'A': [], 'B': [], 'C': [], 'D': [], 'E': [], 'F': [], 'G': [], 'H': [], 'I': [], 'J': [], 'K': [], 'L': [], 'M': [], 'N': [], 'O': [], 'P': [], 'Q': [], 'R': [], 'S': [], 'T': [], 'U': [], 'V': [], 'W': [], 'X': [], 'Y': [], 'Z': []}

**Q5.g.ii** Remplacer chaque valeur du dictionnaire `Dlettres` par le codage de la clé correspondante sous la forme d'une liste de chaînes de caractères.  
Par exemple, la valeur de `Dlettres['A']` sera la liste `[' xxx ', 'x   x', 'x   x', 'xxxxx', 'x   x', 'x   x', 'x   x']`.

***Q5.g.ii corrigé***

On reprend la remarque faite en corrigé de la question **Q5.f**, selon laquelle, sur chaque ligne, `L[i]`, $0\leqslant i \leqslant 5$, la $(n+1)^{\rm ème}$ lettre de l'alphabet est codée par la tranche ${\rm L}[i][6n: 6n + 5]$. 

In [52]:
for n in range(26):
    lettre = majuscules[n]
    Dlettres[lettre] = [ligne[6 * n:6 * n + 5] for ligne in L]

In [53]:
## Vérification
Dlettres['Z']

['xxxxx', '    x', '   x ', '  x  ', ' x   ', 'x    ', 'xxxxx']

**[préliminaire pour la question Q6]** On souhaite utiliser le dictionnaire `Dlettres` pour afficher un texte quelconque.  
Il s'agira, partant d'une chaîne, `texte`, composée uniquement des 26 lettres majuscules de l'alphabet et d'espaces, de construire une liste de 7 lignes composées de la concaténation des sept lignes correspondant à chaque lettre du texte.  
Les espaces dans le texte d'origine seront codées par sept lignes comportant une espace, sous la forme d'une liste nommée `espace`.  

**Q6.a** Rappeler ci-dessous, comment, par une compréhension de listes, on peut obtenir la liste des caractères d'une chaîne `ch`.

In [54]:
ch = 'abracadabra'
liste_caracteres = [...]

***Q6.a corrigé***

In [55]:
ch = 'abracadabra'
liste_caracteres = [c for c in ch] # ou bien [ch[k] for k in range(len(ch))]

In [56]:
# vérification
liste_caracteres

['a', 'b', 'r', 'a', 'c', 'a', 'd', 'a', 'b', 'r', 'a']

**Q6.b** Définir, à l'aide d'une compréhension de listes, la liste `espace`.  
Définir une copie, `Dcaracteres`, du dictionnaire `Dlettres`, et ajouter à cette copie la clé `' '` associée à la liste `espace`.

In [57]:
espace = [...]
...

Ellipsis

***Q6.b corrigé***

In [58]:
espace = [' ' for i in range(7)]

In [59]:
# vérification
espace

[' ', ' ', ' ', ' ', ' ', ' ', ' ']

In [60]:
Dcaracteres = Dlettres.copy()
Dcaracteres[' '] = espace

In [61]:
# Vérification
Dcaracteres

{'A': [' xxx ', 'x   x', 'x   x', 'xxxxx', 'x   x', 'x   x', 'x   x'], 'B': ['xxxx ', 'x   x', 'x   x', 'xxxx ', 'x   x', 'x   x', 'xxxx '], 'C': [' xxx ', 'x   x', 'x    ', 'x    ', 'x    ', 'x   x', ' xxx '], 'D': ['xxxx ', 'x   x', 'x   x', 'x   x', 'x   x', 'x   x', 'xxxx '], 'E': ['xxxxx', 'x    ', 'x    ', 'xxxx ', 'x    ', 'x    ', 'xxxxx'], 'F': ['xxxxx', 'x    ', 'x    ', 'xxxx ', 'x    ', 'x    ', 'x    '], 'G': [' xxx ', 'x   x', 'x    ', 'x    ', 'x  xx', 'x   x', ' xxx '], 'H': ['x   x', 'x   x', 'x   x', 'xxxxx', 'x   x', 'x   x', 'x   x'], 'I': [' xxx ', '  x  ', '  x  ', '  x  ', '  x  ', '  x  ', ' xxx '], 'J': ['  xxx', '   x ', '   x ', '   x ', '   x ', 'x  x ', ' xx  '], 'K': ['x   x', 'x  x ', 'x x  ', 'xx   ', 'x x  ', 'x  x ', 'x   x'], 'L': ['x    ', 'x    ', 'x    ', 'x    ', 'x    ', 'x    ', 'xxxxx'], 'M': ['x   x', 'xx xx', 'x x x', 'x x x', 'x   x', 'x   x', 'x   x'], 'N': ['x   x', 'x   x', 'xx  x', 'x x x', 'x  xx', 'x   x', 'x   x'], 'O': [' xxx ', 'x  

**Q6.c** On considère une liste `L` de $n$ listes composées chacune de $p$ chaînes de caratères.  
Par exemple, pour $n=3$ et $p=2$ :

In [62]:
L = [['abc', 'jkl'], [' ', ' '], ['def' ,'mno'], [' ', ' '], ['ghi', 'pqr']]

Construire, à l'aide de deux boucles imbriquée, en complétant la cellule ci-dessous, la liste `Lc`, dont les éléments sont les $p$ chaînes de caractères obtenues par concaténation des chaînes situées au même positions dans chacune des sous-listes de `L`.  
Sur l'exemple, on devra obtenir la liste `['abc def ghi', 'jkl mno pqr']`.

In [63]:
L = [['abc', 'jkl'], [' ', ' '], ['def' ,'mno'], [' ', ' '], ['ghi', 'pqr']]

n, p = len(L), len(L[0])
Lc = p * ['']
for j in range(p):  ## pour chaque valeur de j, on va construire la chaîne d'indice j de Lc
    for i in range(n): ## on va ajouter à Lc[j] la chaîne d'indice j de chacune des sous-listes L[i]
        Lc[j] = ... # on ajoute à Lc[j] la chaîne L[i][j]

***Q6.c corrigé***

In [64]:
L = [['abc', 'jkl'], [' ', ' '], ['def' ,'mno'], [' ', ' '], ['ghi', 'pqr']]

n, p = len(L), len(L[0])
Lc = p * ['']
for j in range(p):  ## pour chaque valeur de j, on va construire la chaîne d'indice j de Lc
    for i in range(n): ## on va ajouter à Lc[j] la chaîne d'indice j de chacune des sous-listes L[i]
        Lc[j] = Lc[j] + L[i][j] # on ajoute à Lc[j] la chaîne L[i][j]

In [65]:
# vérification
Lc

['abc def ghi', 'jkl mno pqr']

**Q6.d** On considère une liste de listes `L`. À l'aide d'une boucle, construire une copie, `Lb` de `L`, dans laquelle on aura inséré entre chaque sous-liste de `L` une même liste `L0`.  
Par exemple, à partir des listes `L = [['abc', 'jkl'], ['def' ,'mno'], ['ghi', 'pqr']]` et `L0 = [' ', ' ']`, on devra obtenir la liste `[['abc', 'jkl'], [' ', ' '], ['def' ,'mno'], [' ', ' '], ['ghi', 'pqr']]`.

In [66]:
L = [['abc', 'jkl'], ['def' ,'mno'], ['ghi', 'pqr']]
L0 = [' ', ' ']

Lb = []
...

Ellipsis

***Q6.d corrigé***

In [67]:
L = [['abc', 'jkl'], ['def' ,'mno'], ['ghi', 'pqr']]
L0 = [' ', ' ']

Lb = []
for i in range(len(L) - 1):
    Lb.append(L[i])
    Lb.append(L0)
Lb.append(L[len(L) - 1])

In [68]:
# vérification
Lb

[['abc', 'jkl'], [' ', ' '], ['def', 'mno'], [' ', ' '], ['ghi', 'pqr']]

**Q7** En reprenant les réponses aux questions **Q6.a**, **Q6.b**, **Q6.c**, **Q6.d** et **Q5.e**, définir une fonction `compose_texte(texte)`, renvoyant une unique chaîne de caractères, permettant d'afficher, à l'aide de la fonction `print`, le texte `texte` passé en argument, avec la police de caractères définie en introduction.  
Chaque caractère de la chaîne `texte` sera, dans l'affichage, séparée de sa précédente par une espace.  
L'instruction `print(compose_texte('GUIDO VON ROSSUM'))` devra afficher :  
<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_GUIDO%20VON%20ROSSUM.png"/>
*Note :* Guido Von Rossum est le nom de l'inventeur du langage Python.

On suivra le squelette de fonction suivant :

In [69]:
def compose_texte(texte):
    liste_caracteres = [...] ## <-- construction de la liste des caractères composant le texte (cf. Q6.a)
    ## construction de la liste L composée des listes de listes, définies dans le dictionnaire Dcaracteres (Q6.b)
    ## correspondant à chaque caractère de la liste liste_caracteres
    L = [...] ## on utilisera une compréhension de liste
    ## ajout dans une nouvelle liste L2, entre chaque sous-liste de L, de la liste de listes codant l'espace (cf. Q6.d)
    ## (cf. )
    L2 = []
    ...
    ## construction d'une liste Lt, dont chacun des 7 éléments est la concaténation des sous-listes de même indice
    ## correspondant à chaque caractère (cf. Q6.c)
    Lt = ...
    for j in range(7):
        for i in range(len(L2)):
            Lt[j] = ...
    ## concaténation des 7 chaînes de caractères de la liste Lt, séparées par des sauts de ligne, '\n', 
    ## afin d'obtenir sous la forme d'une unique chaîne de caractères, ch, le texte à afficher (cf. Q5.e)
    ch = ''
    for j in range(len(Lt) - 1):
        ch = ...
    ch = ...
    ##
    return ch

***Q7 corrigé***

In [70]:
def compose_texte(texte):
    liste_caracteres = [c for c in texte] ## <-- construction de la liste des caractères composant le texte (cf. Q6.a)
    ## construction de la liste L composée des listes de listes, définies dans le dictionnaire Dcaracteres (Q6.b)
    ## correspondant à chaque caractère de la liste liste_caracteres
    L = [Dcaracteres[c] for c in liste_caracteres] ## on utilisera une compréhension de liste
    ## ajout dans une nouvelle liste L2, entre chaque sous-liste de L, de la liste de listes codant l'espace (cf. Q6.d)
    ## (cf. )
    L2 = []
    for i in range(len(L) - 1):
        L2.append(L[i])
        L2.append(espace)
    L2.append(L[len(L) - 1])
    ## construction d'une liste Lt, dont chacun des 7 éléments est la concaténation des sous-listes de même indice
    ## correspondant à chaque caractère (cf. Q6.c)
    Lt = 7 * ['']
    for j in range(7):
        for i in range(len(L2)):
            Lt[j] = Lt[j] + L2[i][j]
    ## concaténation des 7 chaînes de caractères de la liste Lt, séparées par des sauts de ligne, '\n', 
    ## afin d'obtenir sous la forme d'une unique chaîne de caractères, ch, le texte à afficher (cf. Q5.e)
    ch = ''
    for j in range(len(Lt) - 1):
        ch = ch + Lt[j] + '\n'
    ch = ch + Lt[len(Lt) - 1]
    ##
    return ch

* Traitement de l'exemple

In [71]:
print(compose_texte('GUIDO VON ROSSUM'))

 xxx  x   x  xxx  xxxx   xxx    x   x  xxx  x   x   xxxx   xxx   xxxx  xxxx x   x x   x
x   x x   x   x   x   x x   x   x   x x   x x   x   x   x x   x x     x     x   x xx xx
x     x   x   x   x   x x   x   x   x x   x xx  x   x   x x   x x     x     x   x x x x
x     x   x   x   x   x x   x   x   x x   x x x x   xxxx  x   x  xxx   xxx  x   x x x x
x  xx x   x   x   x   x x   x    x x  x   x x  xx   x x   x   x     x     x x   x x   x
x   x x   x   x   x   x x   x    x x  x   x x   x   x  x  x   x     x     x x   x x   x
 xxx   xxx   xxx  xxxx   xxx      x    xxx  x   x   x   x  xxx  xxxx  xxxx   xxx  x   x


## Exercice 2 encodage d'une image sous la forme d'un fichier `.pbm`

On considère l'image de pomme présentée en introduction :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_pomme_NB.png" width="100"/>

On définit ici un encodage usuel pour une image en noir et blanc, dans lequel la couleur noire est codée par l'entier $0$ et la couleur blanche par l'entier $1$.  

On donne ici l'encodage partiel de l'image sous la forme d'un tableau d'entiers à deux dimensions, de taille $16\times 16$ :

In [72]:
image_nb = [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1], \
            [1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1], \
            [1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]

**L'<span style="color:red">encodage est partiel</span> car les deux premières lignes de la matrice, ainsi que la dernière, ont tous leurs pixels à $1$**.

On peut obtenir l'affichage dans Jupyter (ou tout autre IDE) de l'image ainsi codée, en appelant la fonction suivante (dont le code est donné, mais n'est pas à connaître) :

In [73]:
def affiche_image_nb(tab):
    # https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.figure.html#
    import matplotlib.pyplot as plt
    fig = plt.figure()
    plt.imshow(tab, cmap = "gray")
    plt.show()

* Affichage de l'image codée par un tableau de zéros et de uns, ici le tableau `image_nb` :

In [74]:
affiche_image_nb(image_nb) ## image altérée de la pomme

On rappelle la numérotation des lignes et des colonnes pour un tel tableau (ici sur l'image non altérée de la pomme) :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_pomme_NB_nos_lignes_colonnes.png" width="300"/>

**Q1** Exécuter la cellule ci-dessous pour être sûr de disposer du tableau `image_nb` dans son état dorigine.  
Puis, afficher la valeur du pixel `image_nb[1][8]`. puis changer sa valeur, à l'aide d'une instruction d'affectation :  `image_nb[1][8] = ...`, de sorte que sa couleur corresponde à celle de l'image non altérée de la pomme)

In [75]:
image_nb = [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1], \
            [1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1], \
            [1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]

***Q1 corrigé***

In [76]:
image_nb = [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1], \
            [1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1], \
            [1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]

In [77]:
image_nb[1][8]

1

In [78]:
image_nb[1][8] = 0

**Q2** Modifier de même les valeurs des neuf autres pixels dont la couleur ne correspond pas à l'image originale, de sorte que le tableau `image_nb` code bien l'image originale.  
***On pourra lister les coordonnées, (ligne, colonne), des pixels à modifier et utiliser une boucle `for`***

* Vérifier le résultat en appelant la fonction `affiche_image_nb`

***Q2 corrigé***

In [79]:
liste_pixel_a_modifier = [(0,9), (0,10), (1,8), (1,11), (15,4), (15,5), (15,6), (15,8), (15,9), (15,10)]
for pixel in liste_pixel_a_modifier:
    image_nb[pixel[0]][pixel[1]] = 0

In [80]:
affiche_image_nb(image_nb)

**Q3** On veut définir l'encodage sous ce format de la lettre &laquo; O &raquo; selon le schéma présenté en introduction de l'exercice 1 :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_imageNB_lettreO.png"/>

Définir l'encodage de cette lettre &laquo; O &raquo; sous la forme du tableau $7\times 5$, `image_O` de zéros et de un.  
***On devra utiliser <u>uniquement</u> les opérations de concaténation et de duplication, `+` et*** `*` ***pour les listes***.  
***On construira d'abord les deux listes, `liste0` et `liste1`, codant les deux premières lignes, et on les utilisera pour composer la liste de liste `liste_O`***.

Afficher la lettre à l'aide de la fonction `affiche_lettre_nb`.

**Q3 corrigé**

Une construction élémentaire serait la suivante :

In [81]:
image_O = [[1] + 3 * [0] + [1], [0] + 3 * [1] + [0], [0] + 3 * [1] + [0], [0] + 3 * [1] + [0], \
           [0] + 3 * [1] + [0], [0] + 3 * [1] + [0], [1] + 3 * [0] + [1]]

Si on suit les instructions, on fera :

In [82]:
ligne0 = [1] + 3 * [0] + [1]
ligne1 = [0] + 3 * [1] + [0]

In [83]:
ligne0, ligne1

([1, 0, 0, 0, 1], [0, 1, 1, 1, 0])

et on construira alors la liste voulue de la façon suivante :

In [84]:
image_O = [ligne0] + 5 * [ligne1] + [ligne0]

In [85]:
image_O

[[1, 0, 0, 0, 1], [0, 1, 1, 1, 0], [0, 1, 1, 1, 0], [0, 1, 1, 1, 0], [0, 1, 1, 1, 0], [0, 1, 1, 1, 0], [1, 0, 0, 0, 1]]

In [86]:
affiche_image_nb(image_O)

**Q4** On veut définir l'encodage sous ce format de la lettre &laquo; K &raquo; selon le schéma présenté en introduction de l'exercice 1 :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_imageNB_lettreK.png"/>

**Q4.a** Définir un tableau $7\times 5$, `image_K` rempli de uns (on modifiera ensuite les valeurs des pixels *ad hoc* pour obtenir la lettre 'K'.  

**On prêtera attention au fait que les lignes de 1 construites devront être indépendantes (ne pas être des copies d'une même ligne).**  

<span style="color:red">***Il faudra donc construire sept lignes de cinq 1, de sorte que la modification de l'une de ces lignes n'affecte pas les autres.***</span> On privilégiera d'utiliser une compréhension de listes.

*Note : si on demande l'affichage de `image_K` à ce stade, on obtiendra une image noire, mais ceci est un bug d'affichage.* Si on examine le contenu de la variable `image_K`, on pourra constater qu'il s'agit bien d'un tableau d'entiers égaux à 1.

***Q4.a corrigé***

On peut construire le tableau demandé à partir d'une liste vide, par ajouts successifs :

In [87]:
image_K = []
for i in range(7):
    image_K.append(5 * [1])

In [88]:
# vérification
image_K

[[1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1]]

Ou utiliser une compréhension de liste (ce qui est spécifiquement demandé) :

In [89]:
image_K = [5 * [1] for i in range(7)]

In [90]:
# vérification
image_K

[[1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1]]

*Note :* à ce stade, l'affichage de l'image codée par le tableau `image_K` donne un rectangle noir (alors que l'on attend un rectangle blanc) : ceci est un bug d'affichage, qui disparaît dès qu'il y a au moins un pixel à 0 dans la matrice).

In [91]:
affiche_image_nb(image_K)

**Q4.b** À l'aide de trois boucles `for` (une par élément) :  
* colorier la barre verticale à gauche de la lettre 'K' ;  
* colorier chaque barre oblique de la lettre 'K'.

Afficher la lettre à l'aide de la fonction `affiche_image_nb`.

***Q4.b corrigé***

In [92]:
image_K = [5 * [1] for i in range(7)]

* coloriage de la barre verticale (pixels d'indices $(0, 0), (1, 0), \ldots, (6, 0)$) :

In [93]:
for i in range(7):
    image_K[i][0] = 0

In [94]:
affiche_image_nb(image_K)

* coloriage de la barre oblique montante (pixels d'indices $(3, 1), (2, 2), (1, 3), (0, 4)$) :

On remarque que l'indice de colonne, que l'on notera $j$, varie de 1 à 4, tandis que l'indice de ligne peut-être calculé par la formule $i=4 - j$.

In [95]:
for j in range(1, 5):
    image_K[4 - j][j] = 0

In [96]:
affiche_image_nb(image_K)

* coloriage de la barre oblique descendante (pixels d'indices $(4, 2), (5, 3), (6, 4)$) :

On remarque, par exemple, que l'indice de colonne, que l'on notera $j$, varie de 2 à 4, tandis que l'indice de ligne peut-être calculé par la formule $i=2 + j$.

In [97]:
for j in range(2, 5):
    image_K[2 +  j][j] = 0

In [98]:
affiche_image_nb(image_K)

**Q5** On veut définir l'encodage sous ce format de la lettre &laquo; Z &raquo; selon le schéma présenté en introduction de l'exercice 1 :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_imageNB_lettreZ.png"/>

En utilisant compréhensions de listes et concaténation de listes, créer un tableau $7\times 5$ `image_Z` dont la première et dernière lignes ne contiennent que des zéros et les autres que des uns.  
Utiliser ensuite une boucle `for` pour mettre à zéro les pixels correspondant à la barre oblique du Z.

Afficher la lettre à l'aide de la fonction `affiche_lettre_nb`.

***Q5 corrigé***

In [99]:
image_Z = [[0 for j in range(5)]] + [[1 for j in range(5)] for k in range(5)] + [[0 for j in range(5)]]

On a ainsi construit le tableau qu'il a été demandé de construire préliminairement :

In [100]:
image_Z

[[0, 0, 0, 0, 0], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [0, 0, 0, 0, 0]]

Pour *colorier* la barre oblique du Z, on peut remarquer que, dans les lignes d'indice $i$, $1\leqslant i\leqslant 5$, seul le pixel d'indice $j=5-i$ est à mettre à zéro.

In [101]:
for i in range(1, 6):
    image_Z[i][5 - i] = 0

In [102]:
affiche_image_nb(image_Z)

***\[ne pas faire ce qui suit - fin de l'exercice 2 - pendant le temps du TP: <span style="background-color:yellow;color:red"> passer à l'exercice 4</span>\]***

**[prélimaire pour la question Q6]** Il est possible d'enregistrer une image en noir et blanc dans un fichier texte, sous un format proche de la représentation précédente, le <a href="https://fr.wikipedia.org/wiki/Portable_pixmap">format `.pbm`</a> (pour *portable bitmap*).  
Avec ce format, l'encodage de la lettre &laquo; A &raquo; selon le schéma de l'introduction, sera réalisé par le fichier suivant :<br/>
<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_lettreA-pbm.png"/> 

* La première ligne de ce fichier indique le format d'encodage ('P1' pour un fichier en noir et blanc).  
* La seconde ligne est une ligne de commentaire (libre) et doit débuter par un `#`.  
* La troisième ligne indique, dans cet ordre, le nombre de colonnes de l'image et son nombre de lignes.  
* Les lignes suivantes donnent les valeurs de chaque pixel de l'image (ici ligne par ligne), séparées par par un symbole (*séparateur*) conventionnel qui peut être un espace, une tabulation ou un saut de ligne.<br/>  

*Note : le nombre de caractères par ligne ne doit pas excéder $70$ caractères*.

**Q6** On reprend ici l'encodage de la lettre &laquo; A &raquo; défini dans le préliminaire :

In [103]:
A = " xxx \nx   x\nx   x\nxxxxx\nx   x\nx   x\nx   x\n"

**Q6.a** Construire, à l'aide d'une boucle `for`, une chaîne de caractères `A01` dans laquelle les espaces auront été remplacés par le chiffre `'1'` et les caractères `'x'` auront été remplacés par le chiffre `'0'`.

On pourra utiliser un branchement conditionnel

In [104]:
if ... :    ## on teste ici si une condition 1 est réalisée
    ...     ## on effectue ici le traitement 1 prévu si la condition 1 est réalisée
            ## on sort alors du branchement
elif ...:   ## on teste ici si une condition 2 est réalisée
    ...     ## on effectue ici le traitement 2 prévu si la condition 2 est réalisée
            ## on sort alors du branchement
else:
    ...     ## on effectue ici traitement 3 prévu si ni la condition 1,ni la condition 2 ne sont réalisées
            ## on sort alors du branchement

équivalente à la construction

In [105]:
if ...:
    ...
else:
    if ...:
        ...
    else:
        ...

***Q6.a corrigé***

In [106]:
A01 = ""
for car in A:
    if car == ' ':
        A01 = A01 + '1'
    elif car == 'x':
        A01 = A01 + '0'
    else:
        A01 = A01 + car

In [107]:
# vérification
A01

'10001\n01110\n01110\n00000\n01110\n01110\n01110\n'

**Q6.b** Afin de se conformer au format `.pbm`, construire, à l'aide d'une boucle `while`, une chaîne de caractères `Apbm`, dans laquelle les chiffres dans chaque ligne auront été séparés par des espaces.  
On devra obtenir la chaîne `'1 0 0 0 1\n0 1 1 1 0\n0 1 1 1 0\n0 0 0 0 0\n0 1 1 1 0\n0 1 1 1 0\n0 1 1 1 0\n'`

***Q6.b corrigé***

Il s'agit de remplacer chaque caractère, `'0'` ou `'1'`, par, respectivement `'0 '` ou `'1 '`, sauf si le caractère est suivi par un caractère de passage à la ligne.

In [108]:
Apbm = ""
i = 0
while i < len(A01):
    if A01[i+1] == '\n':          ## si le caractère lu est suivi d'un \n,
        Apbm = Apbm + A01[i:i+2]  ## on copie le caractère ainsi que le caractère \n qui le suit
        i = i + 2                 ## on avance de deux rangs dans la chaîne pour traiter le caractère, 0 ou 1, suivant
    else:                             ## si le caractère lu n'est pas suivi d'un \n,
        Apbm = Apbm + A01[i] + ' '    ## on copie le caractère en le faisant suivre d'une espace
        i = i + 1                     ## on avance d'un rang dans la chaîne pour traiter le caractère, 0 ou 1, suivant

In [109]:
# Vérification
Apbm

'1 0 0 0 1\n0 1 1 1 0\n0 1 1 1 0\n0 0 0 0 0\n0 1 1 1 0\n0 1 1 1 0\n0 1 1 1 0\n'

**[prélimaire pour la question Q7]** On redonne ici une façon de créer un fichier texte dans le répertoire de travail courant :

In [110]:
with open('myfile.ext', 'w') as f:
    f.write('abcdefghijklm\n')
    f.write('nopqrst')
    f.write('uvwxyz')

On a défini ici le fichier suivant :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_edition_myfile-ext.png"/>

Chaque instruction `f.write(ch)` permet d'ajouter, à la fin du fichier, la chaîne de caractères, `ch`, passée en argument.  
Les passages à la ligne ne sont enregistrés que si l'on écrit explicitement un caractère de fin de ligne, `'\n'` dans le fichier.

Pour ouvrir un tel fichier (dont l'extension n'est pas l'extension par défaut `.txt`), avec un éditeur de texte, il faut d'abord ouvrir l'éditeur, puis cliquer sur &laquo; Ouvrir &raquo;, parcourir l'arborescence des fichiers pour aller dans le dossier contenant le fichier, et sélectionner &laquo; Tous les fichiers &raquo; pour que le nom du fichier apparaisse
<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/ouverture%20fichier%20.pbm%20avec%20bloc%20notes-1.png"/>
<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/ouverture%20fichier%20.pbm%20avec%20bloc%20notes-2.png"/>

**Q7** Compléter le code de la fonction suivante, `sauve_pixmap`, afin de pouvoir sauvegarder une image dans un fichier texte au format `.pbm` en respectant ses spécifications.

In [111]:
def sauve_pixmap(chemin_vers_fichier, formatpixmap, commentaire, nbcolonnes, nblignes, image):
    with open(..., ...) as f:
        f.write(formatpixmap + '\n')       ## écriture de la première ligne
        f.write(...)                       ## écriture de la deuxième ligne
        f.write(...)                       ## écriture de la troisième ligne
        f.write(...)                       ## écriture de la chaîne de caractère encodant l'image

L'instruction `sauve_pixmap('malettreA.pbm', 'P1', 'Un exemple bitmap de la lettre "A"', 5, 7, Apbm)` devra créer, dans le répertoire courant du notebook, un fichier exactement identique au fichier présenté en préliminaire de la **question 6**.

Comme la méthode `write` ne peut prendre en argument qu'une valeur de type `str`, on rappelle, pour la question **Q6**, l'usage de la fonction de conversion de type `str()` :

In [112]:
str(12)

'12'

***Q7 corrigé***

In [113]:
def sauve_pixmap(chemin_vers_fichier, formatpixmap, commentaire, nbcolonnes, nblignes, image):
    with open(chemin_vers_fichier, 'w') as f:
        f.write(formatpixmap + '\n')       ## écriture de la première ligne
        f.write('#' + commentaire + '\n')  ## écriture de la deuxième ligne
        f.write(str(nbcolonnes) + ' ' + str(nblignes) + '\n')  ## écriture de la troisième ligne
        f.write(image)                     ## écriture de la chaîne de caractère encodant l'image

* test

In [114]:
sauve_pixmap('malettreA.pbm', 'P1', 'Un exemple bitmap de la lettre "A"', 5, 7, Apbm)

In [115]:
# vérification : on ouvre le fichier pour examiner son contenu
f = open('malettreA.pbm')
contenu_fichier = f.read()
f.close()
# affichage du contenu du fichier
print(contenu_fichier)

P1
#Un exemple bitmap de la lettre "A"
5 7
1 0 0 0 1
0 1 1 1 0
0 1 1 1 0
0 0 0 0 0
0 1 1 1 0
0 1 1 1 0
0 1 1 1 0



## Exercice 3 Lecture d'un fichier `.pbm`

On a téléchargé en début de ce notebook un fichier `pomme.pbm` correspondant à l'encodage au format `.pbm` de l'image de pomme suivante :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_pomme_NB.png" width="100"/>

**[prélimaire pour la question Q1]** On redonne ici une façon de lire un fichier texte, `haiku.txt`, présent dans le travail courant, comportant trois lignes, où le fichier est lu ligne par ligne, à l'aide de la méthode `.readline()` qui lit les lignes du fichier, les unes après les autres, à chaque appel à la méthode.

In [116]:
f = open('haiku.txt')
ligne1 = f.readline()
ligne2 = f.readline()
ligne3 = f.readline()
finfichier = f.readline()
f.close()

Voici les lignes lues :

In [117]:
ligne1, ligne2, ligne3

('Un vieil étang\n', 'Une grenouille qui plonge,\n', "Le bruit de l'eau.")

***Lorsque l'on atteint la fin du fichier, la méthode `.readline()` renvoie une chaîne vide, ce qui signale que la fin du fichier a été atteinte.***   
Ainsi la chaîne de caractères, `finfichier`, lue par lors du dernier appel à la méthode `.readline()` dans l'avant-dernière cellule ci-avant, est une chaîne vide :

In [118]:
finfichier

''

La lecture d'une ligne vide, elle, renverrait la chaîne `\n`, ce qui permet de différencier sans ambigïté la fin de fichier et une ligne vide dans le fichier.  
Ainsi, on peut lire l'intégralité d'un fichier à l'aide d'une boucle `while` et de la méthode `.readline()`, en stockant chaque ligne lue dans une liste `lignes_fichier`, comme ci-dessous :

In [119]:
f = open('haiku.txt', 'r') # l'argument 'r' (pour read) peut-être omis
lignes_fichier = []
lignelue = f.readline()
while lignelue != '':
    lignes_fichier.append(lignelue)
    lignelue = f.readline()
f.close()

Voici la liste de chaînes de caractères ainsi obtenue :

In [120]:
lignes_fichier

['Un vieil étang\n', 'Une grenouille qui plonge,\n', "Le bruit de l'eau."]

**Q1.a** Ouvrir et lire à l'aide de la dernière procédure décrite ci-avant le fichier `pomme.pbm` contenant l'encodage de la pomme. *On stockera les lignes lues dans une liste `lignes_pbm_pomme`*.

***Q1.a corrigé***

In [121]:
f = open('pomme.pbm', 'r')
lignes_pbm_pomme = []
lignelue = f.readline()
while lignelue != '':
    lignes_pbm_pomme.append(lignelue)
    lignelue = f.readline()
f.close()

In [122]:
# vérification
lignes_pbm_pomme

['P1\n', '# une image de pomme en noir et blanc\n', '16 16\n', '1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1\n', '1 1 1 1 1 1 1 1 0 1 1 0 1 1 1 1\n', '1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1\n', '1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1\n', '1 1 1 1 0 0 1 0 1 0 0 0 1 1 1 1\n', '1 1 1 0 1 0 0 0 0 0 1 1 0 1 1 1\n', '1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1\n', '1 0 0 1 1 1 1 1 1 1 1 1 1 0 1 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1\n', '1 1 0 1 1 1 1 1 1 1 1 1 0 1 1 1\n', '1 1 1 0 1 1 0 0 0 1 1 0 1 1 1 1\n', '1 1 1 1 0 0 0 1 0 0 0 1 1 1 1 1\n']

**Q1.b** Extraire de la liste `lignes_pbm_pomme`, une liste, `image_pbm_pomme`, des 16 lignes encodant le dessin de la pomme. On utilisera le *slicing*.

***Q1.b corrigé***

Il s'agit d'extraire de la liste `lignes_pbm_pomme`, les sous-listes d'indices $3$ à ${\rm len(lignes\_pbm\_pomme)} -1$ :

In [123]:
lignes_pbm_pomme = lignes_pbm_pomme[3: len(lignes_pbm_pomme)]

In [124]:
# vérification
lignes_pbm_pomme

['1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1\n', '1 1 1 1 1 1 1 1 0 1 1 0 1 1 1 1\n', '1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1\n', '1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1\n', '1 1 1 1 0 0 1 0 1 0 0 0 1 1 1 1\n', '1 1 1 0 1 0 0 0 0 0 1 1 0 1 1 1\n', '1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1\n', '1 0 0 1 1 1 1 1 1 1 1 1 1 0 1 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1\n', '1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1\n', '1 1 0 1 1 1 1 1 1 1 1 1 0 1 1 1\n', '1 1 1 0 1 1 0 0 0 1 1 0 1 1 1 1\n', '1 1 1 1 0 0 0 1 0 0 0 1 1 1 1 1\n']

**Q1.c** Construire, à partir de la liste `lignes_pbm_pomme`, un tableau d'entiers, `image_pomme`, de taille $16\times 16$, dont les éléments, égaux à $0$ ou $1$, encodent l'image en noir et blanc de la pomme.  
**On parcourra chaque élément de la liste `lignes_pbm_pomme` pour en lire uniquement les caractères d'indice pair (qui seuls correspondent aux chiffres '0' et '1').**  On pourra utiliser pour cela indifféremment une boucle `for` ou `while`.  
On convertira en entier les chiffres lus, `'0'` ou `'1'`, à l'aide de la fonction de conversion de type `int` :

In [125]:
int('0'), int('1')

(0, 1)

On utilisera le squelette suivant :

In [126]:
image_pomme = []
for ligne in lignes_pbm_pomme:
    image_pomme.append([]) ## on crée une nouvelle sous-liste dans la liste image_pomme, cette sous-liste 
                           ## à pour indice i dans la liste image_pomme, il s'agit de image_pomme[i]
    ## parcours de la chaîne ligne pour en extraire les chiffres '0' '1' et les ajouter, un par un,
    ## à la sous-liste image_pomme[i]
    
    

Afficher l'image de la pomme à l'aide de la fonction `affiche_image_nb`.

***Q1.c corrigé***

* en utilisant une boucle `while` pour le parcours de chaque sous-liste `ligne`

In [127]:
image_pomme = []
for i in range(len(lignes_pbm_pomme)):
    image_pomme.append([]) ## on crée une nouvelle sous-liste dans la liste image_pomme, cette sous-liste 
                           ## à pour indice i dans la liste image_pomme, il s'agit de image_pomme[i]
    ## parcours de la chaîne lignes_pbm_pomme[i] pour en extraire les chiffres '0' '1' et les ajouter, un par un,
    ## à la sous-liste image_pomme[i], à l'aide de la méthode .append
    ligne = lignes_pbm_pomme[i]
    j = 0
    while j < len(ligne):
        image_pomme[i].append(int(ligne[j]))
        j = j + 2

In [128]:
# vérification
image_pomme

[[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1], [1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1], [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], [1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1], [1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1], [1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1]]

In [129]:
affiche_image_nb(image_pomme)

* en utilisant une boucle `for` pour le parcours de chaque sous-liste `ligne`

In [130]:
image_pomme = []
for i in range(len(lignes_pbm_pomme)):
    image_pomme.append([]) ## on crée une nouvelle sous-liste dans la liste image_pomme, cette sous-liste 
                           ## à pour indice i dans la liste image_pomme, il s'agit de image_pomme[i]
    ## parcours de la chaîne lignes_pbm_pomme[i] pour en extraire les chiffres '0' '1' et les ajouter, un par un,
    ## à la sous-liste image_pomme[i], à l'aide de la méthode .append
    ligne = lignes_pbm_pomme[i]
    for ip in range(len(ligne) // 2):
        image_pomme[i].append(int(ligne[2 * ip]))

In [131]:
# vérification
image_pomme

[[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1], [1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1], [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], [1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1], [1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1], [1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1]]

In [132]:
affiche_image_nb(image_pomme)

In [133]:
pomme = [[0 for j in range(16)] for i in range(16)]
for i in range(len(lignes_pbm_pomme)):
    for j in range (0,len(lignes_pbm_pomme[i]),2):
        pomme[i][j//2] = int(lignes_pbm_pomme[i][j])

pomme2 = []
for i in range(len(lignes_pbm_pomme)):
    pomme2.append([])
    for j in range (0,len(lignes_pbm_pomme[i]),2):
        pomme2[i].append(int(lignes_pbm_pomme[i][j]))

In [134]:
affiche_image_nb(pomme)
affiche_image_nb(pomme2)

**[Préliminaire pour la question Q2** Il est possible de combiner les méthodes `.readline()` et `.read()`, pour lire les premières lignes d'un fichier texte, une par une, à l'aide de la méthode `.readline()`, puis l'ensemble de toutes les lignes restantes à l'aide de la méthode `.read()`. On récupère alors l'ensemble des lignes restantes sous la forme d'une unique chaîne de caractères.

On donne aussi pour la question **Q2** les méthodes `.replace(c1, c2)` et `.split` permettant, respectivement :  
* de créer une copie d'une chaîne dans laquelle on a remplacé toutes les occurrences d'un caractère ou d'une chaîne `old` par un caractère ou une chaîne `new`.

In [135]:
"toto".replace('o', 'oo')

'tootoo'

In [136]:
help(str.replace)

Help on method_descriptor:

replace(self, old, new, count=-1, /)
    Return a copy with all occurrences of substring old replaced by new.
    
      count
        Maximum number of occurrences to replace.
        -1 (the default value) means replace all occurrences.
    
    If the optional argument count is given, only the first count occurrences are
    replaced.




* de construire à partir d'une chaîne et d'un séparateur, `sep`, la liste de toutes ses sous-chaînes séparées par le séparateur `sep`.

In [137]:
"aabbab".split('b')

['aa', '', 'a', '']

si aucun séparateur n'est spécifié, c'est le séparateur espace qui est pris en compte :

In [138]:
"guido von rossum".split()

['guido', 'von', 'rossum']

si la chaîne termine par un caractère de passage à la ligne, celui-ci est ignoré :

In [139]:
"guido von rossum\n".split()

['guido', 'von', 'rossum']

In [140]:
help(str.split)

Help on method_descriptor:

split(self, /, sep=None, maxsplit=-1)
    Return a list of the words in the string, using sep as the delimiter string.
    
    sep
      The delimiter according which to split the string.
      None (the default value) means split according to any whitespace,
      and discard empty strings from the result.
    maxsplit
      Maximum number of splits to do.
      -1 (the default value) means no limit.




**Q2** On a téléchargé dans le répertoire courant du notebook, un fichier `lettreG.pbm` correspondant à l'encodage au format `.pbm` de l'image de la lettre &laquo; G &raquo; :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_imageNB_lettreG.png"/>

Le contenu de ce fichier est le suivant :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_lettreG-pbm.png"/>

Comme les méthodes de lecture du contenu d'un fichier texte ne permette de récupérer que des chaînes de caractères, on rappelle l'usage de la fonction de conversion de type `int()`, qui permet de convertir une chaîne représentant un entier en une valeur de type `int` :

In [141]:
int('12')

12

**Q2.a** Ouvrir le fichier `lettreG.pbm` téléchargé, pour en extraire, d'une part, les nombres de colonnes et de lignes de l'image, que l'on stockera dans deux variables entières, respectivement, `p` et `n`, et d'autre part, sous la forme d'une unique chaîne de caractères, `codageG_pbm`, le codage de l'image en noir et blanc de la lettre &laquo; G &raquo;.  
<i><u>Indication</u> :</i> on pourra lire les 3 premières lignes du fichier à l'aide de la méthode `.readline` et toutes les autres lignes à l'aide de la méthode `.read`.

**Q2.a corrigé**

In [142]:
# ouverture de l'accès au fichier en lecture
f = open('lettreG.pbm', 'r')
# lecture des trois premières lignes
ligne1 = f.readline()
ligne2 = f.readline()
ligne3 = f.readline()
# lecture des dernières lignes du fichier
codageG_pbm = f.read()
# fermeture de l'accès au fichier
f.close()

In [143]:
# vérification
ligne1, ligne2, ligne3

('P1\n', '# Un exemple bitmap de la lettre "G"\n', '5 7\n')

In [144]:
codageG_pbm

'1 0 0 0 1 0 1 1 1 0 0 1 1 1 1 0 1 1 \n1 1 0 1 1 0 0 0 1 1 1 0 1 0 0 0 1'

*Note :* les première et deuxième lignes n'étant pas à traiter, on aurait pu ne pas les assigner à une variable et écrire :

In [145]:
f = open('lettreG.pbm', 'r')
for i in range(2):
    f.readline()
ligne3 = f.readline()
# lecture des dernières lignes du fichier
codageG_pbm = f.read()
# fermeture de l'accès au fichier
f.close()

* extraction du nombre de colonnes et du nombre de lignes de l'image :

In [146]:
p, n = ligne3.split()

In [147]:
p, n

('5', '7')

In [148]:
p, n = int(p), int(n) ## conversion du type str au type int

In [149]:
p, n

(5, 7)

**Q2.b** En remarquant bien que l'encodage au format `.pbm` de l'image a été stocké en utilisant seulement deux lignes du fichier texte, construire un tableau d'entiers, `imageG_nb`, de taille $7\times 5$, dont les éléments, égaux à $0$ ou $1$, encodent l'image en noir et blanc de la lettre &laquo; G &raquo;.  
**On traitera la chaîne `codageG_pbm` en une liste de zéros et de uns, `liste_pixels`, en remplaçant le(s) caractère(s) de passage à la ligne `'\n'` par une espace et en utilisant ensuite la méthode `.split`.**  
On utilisera une compréhension de liste pour transformer la liste de chiffres `liste_pixels` en une liste d'entiers.  
On construira ensuite une matrice, `image_G`, de zéros de taille $n\times p$, dont on remplacera les coefficients par les valeurs lues dans des tranches de longueur $p$ de la liste `liste_pixels`.

Afficher l'image de la lettre 'G' l'aide de la fonction `affiche_image_nb`.

**Q2.b corrigé**

In [150]:
codageG_pbm

'1 0 0 0 1 0 1 1 1 0 0 1 1 1 1 0 1 1 \n1 1 0 1 1 0 0 0 1 1 1 0 1 0 0 0 1'

* remplacement du caractère `\n` par une espace et conversion en une liste de chiffres, `'0'` et `'1'` :

In [151]:
liste_pixels = codageG_pbm.replace('\n', ' ').split()

In [152]:
# vérification
liste_pixels

['1', '0', '0', '0', '1', '0', '1', '1', '1', '0', '0', '1', '1', '1', '1', '0', '1', '1', '1', '1', '0', '1', '1', '0', '0', '0', '1', '1', '1', '0', '1', '0', '0', '0', '1']

* conversion de la liste de caractères `liste_pixels` en une liste de chiffres :

In [153]:
liste_pixels = [int(c) for c in liste_pixels]

In [154]:
# vérification
liste_pixels

[1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1]

* création d'une matrice, `image_G`, $n\times p$ de zéros

In [155]:
image_G = [[0 for j in range(p)] for i in range(n)]

In [156]:
# vérification
image_G

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]

* lecture par tranche de longueur $p$ de la liste `liste_pixels` pour recopier les valeurs des pixels dans la matrice `image_G`

In [157]:
for i in range(n):
    tranche = liste_pixels[i * p: i * p + p] ## copie de la (i + 1)-ème tranche
    for j in range(p):   ## copie des valeurs de la tranche dans la ligne d'indice i de la matrice image_G
        image_G[i][j] = tranche[j]

In [158]:
# vérification
image_G

[[1, 0, 0, 0, 1], [0, 1, 1, 1, 0], [0, 1, 1, 1, 1], [0, 1, 1, 1, 1], [0, 1, 1, 0, 0], [0, 1, 1, 1, 0], [1, 0, 0, 0, 1]]

In [159]:
affiche_image_nb(image_G)

Il n'était pas nécessaire de créer une copie de chaque tranche, on pouvait procéder comme suit :

In [160]:
image_G = [[0 for j in range(p)] for i in range(n)]
for i in range(n):
    for j in range(p): 
        image_G[i][j] = liste_pixels[i * p + j] ## copie de la valeur du pixel en ligne i, colonne j
# vérification
affiche_image_nb(image_G)

## Exercice 4 Manipulation d'une image en noir et blanc

On donne ici l'encodage de l'image de pomme (non altérée) utilisée précédemment :

In [161]:
pomme_nb = [[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1], \
            [1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1], \
            [1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], \
            [1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], \
            [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1], \
            [1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1], \
            [1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1]]

In [162]:
affiche_image_nb(pomme_nb)

**[préliminaire à la question Q1.a]**

On rappelle la numérotation des lignes et des colonnes de l'image :

<img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_pomme_NB_nos_lignes_colonnes.png"/>

On rappelle que la copie (profonde) d'une liste de listes, dont les sous-listes sont des listes d'éléments de type non mutable (`int`, `float`, `str`, `bool`, ou tuples d'éléments de l'un de ces quatre types), se fait à l'aide des deux boucles imbriquées : la boucle *principale* parcourant la liste principale, la boucle *intérieure* servant à parcourir chacune des sous-listes :

In [163]:
LL = [[1, 2, 3], [4, 5], [6, 7, 8, 9, 10]]
## copie profonde de la liste de listes LL à partir d'une liste vide
copie_de_LL = []
for i in range(len(LL)):
    copie_de_LL.append([]) ## ajout à la liste de liste copie_de_LL d'une sous-liste d'indice i, vide
    for j in range(len(LL[i])):
        copie_de_LL[i].append(LL[i][j]) ## ajout à la ligne copie_de_LL[i] de la valeur LL[i][j]

* copie (profonde) d'un tableau à deux dimensions dont les éléments de type non mutable (`int`, `float`, `str`, `bool`, ou tuples d'éléments de l'un de ces quatre types) :  

Si la liste de listes est un tableau (toutes les sous-listes ont le même nombre d'éléments et tous les éléments sont de même type), ce qui suit est une adaptation de maniement et de compréhension plus faciles.

In [164]:
tab = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]] ## tableau d'entiers à trois lignes et quatre colonnes
## création d'un tableau de même taille que le tableau à copier, et dont les éléments sont tous égaux (initialement)
copie_de_tab = [[0 for j in range(len(tab[0]))] for i in range(len(tab))]
for i in range(len(tab)):
    for j in range(len(tab[0])):
        copie_de_tab[i][j] = tab[i][j] ## copie en cellule copie_de_tab[i][j] de la valeur tab[i][j]

In [165]:
# résultat
copie_de_tab

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

**copie profonde par compréhensions de listes [abrégé de la méthode précédente] :**

In [166]:
tab = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]] ## tableau d'entiers à trois lignes et quatre colonnes

copie2_de_tab = [[tab[i][j] for j in range(len(tab[0]))] for i in range(len(tab))]

In [167]:
# résultat
copie2_de_tab

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

**Q1** On cherche à obtenir un ***négatif*** de l'image de la pomme.

**Q1.a** Faire une copie (profonde) du tableau `pomme_nb`, que l'on nommera `pomme_nb1`.

***Q1.a corrigé***

In [168]:
pomme_nb1 = [[pomme_nb[i][j] for j in range(len(pomme_nb[0]))] for i in range(len(pomme_nb))]

In [169]:
## vérification
affiche_image_nb(pomme_nb1)

**Q1.b** Modifier le tableau `pomme_nb1` afin de changer les valeurs de chaque pixel : les pixels à 0 passeront à 1, et les pixels à 1 seront mis à zéro.  
**On utilisera deux boucles imbriquées, l'une pour faire varier un indice de ligne $i$, l'autre, intérieure, pour faire varier un indice de colonne $j$.**

Utiliser la fonction `affiche_image_nb` pour afficher l'image encodée désormais par le tableau `pomme_nb1`.  Il s'agit d'un ***négatif*** de l'image d'origine.

***Q1.b corrigé***

In [170]:
for i in range(len(pomme_nb1)):
    for j in range(len(pomme_nb1[0])):
        if pomme_nb1[i][j] == 1:
            pomme_nb1[i][j] = 0
        else:
            pomme_nb1[i][j] = 1

In [171]:
# vérification
affiche_image_nb(pomme_nb1)

ou bien, en utilisant que $1-0=1$ et $1-1=0$ :

In [172]:
pomme_nb1 = [[pomme_nb[i][j] for j in range(len(pomme_nb[0]))] for i in range(len(pomme_nb))]
for i in range(len(pomme_nb1)):
    for j in range(len(pomme_nb1[0])):
        pomme_nb1[i][j] = 1 - pomme_nb[i][j]
# vérification
affiche_image_nb(pomme_nb1)

**Q2** Faire une copie (profonde) du tableau `pomme_nb`, que l'on nommera `pomme_nb2`, puis modifier cette copie de sorte à obtenir le codage de l'image symétrique de l'image de départ dans la symétrie orthogonale ayant pour axe la médiatrice du côté inférieur de l'image.  
**On utilisera que le pixel `pomme_nb[i][j]` a pour symétrique le pixel `pomme_nb[i][p - 1 - j]`, si $p$ est le nombre de colonnes de la matrice-image.**

Utiliser la fonction `affiche_image_nb` pour afficher l'image encodée par le tableau `pomme_nb2`.  Il s'agit d'un *retournement vertical* de l'image d'origine.

***Q2 corrigé***

In [173]:
n, p = len(pomme_nb), len(pomme_nb[0]) ## <-- nombre de lignes, nombre de colonne de la matrice pomme_nb
pomme_nb2 = [[pomme_nb[i][j] for j in range(p)] for i in range(n)]
for i in range(n):
    for j in range(p):
        pomme_nb2[i][j] = pomme_nb[i][p - 1 - j]

In [174]:
# vérification
affiche_image_nb(pomme_nb2)

* Méthode alternative : échanger dans la copie de l'image les pixels symétriques<br/>
**<u>attention</u> :** il ne faut pas poursuivre les échanges au-delà de la colonne médiane, sous peine de revenir à l'image de départ.

In [175]:
n, p = len(pomme_nb), len(pomme_nb[0]) ## <-- nombre de lignes, nombre de colonne de la matrice pomme_nb
pomme_nb2 = [[pomme_nb[i][j] for j in range(p)] for i in range(n)]
for i in range(n):
    for j in range(p // 2): ## <-- on ne parcourt que la partie gauche de l'image pour opérer les échanges
        pomme_nb2[i][j], pomme_nb2[i][p - 1 - j] = pomme_nb2[i][p - 1 - j], pomme_nb2[i][j]
# vérification
affiche_image_nb(pomme_nb2)

**Q3** On cherche à obtenir l'image symétrique de l'image symétrique de l'image de départ dans la symétrie orthogonale ayant pour axe la diagonale *descendante* du carré contenant l'image.  
On commencera par construire une matrice de zéros de taille $p\times n$, où $n$ et $p$ sont, respectivement, les nombres de lignes et de colonnes de l'image de la pomme.  
**On cherchera ensuite quel est le symétrique d'un pixel `pomme_nb[i][j]` dans cette symétrie.**

Utiliser la fonction `affiche_image_nb` pour afficher l'image encodée par le tableau `pomme_nb3`.  Il s'agit d'un *retournement vertical* de l'image d'origine.

**Q3 corrigé**

In [176]:
n, p = len(pomme_nb), len(pomme_nb[0]) ## <-- nombre de lignes, nombre de colonnes de la matrice pomme_nb
pomme_nb3 = [[pomme_nb[i][j] for j in range(n)] for i in range(p)]
for i in range(p):
    for j in range(n):
        pomme_nb3[i][j] = pomme_nb[j][i]

*Note :* on a échangé les nombres de lignes et de colonnes, $n$ et $p$ entre les deux matrices, afin que l'algorithme puisse convenir aussi à une image rectangulaire.

In [177]:
# vérification
affiche_image_nb(pomme_nb3)

* Méthode alternative : échanger dans la copie de l'image les pixels symétriques<br/>
**<u>attention</u> :** il ne faut parcourir que la moitié de l'image pour opérer les échanges : soit la moitié située au-dessus de la diagonale, comme ci-dessous, soit la partie située en-desssous de la diagonale.
**on notera que cette fois la méthode n'est valide que pour une matrice carrée**

In [178]:
n = len(pomme_nb)## <-- nombre de lignes de la matrice-image (qui doit être ici carrée)
pomme_nb3 = [[pomme_nb[i][j] for j in range(n)] for i in range(n)]
for i in range(n):
    for j in range(i + 1, n):
        pomme_nb3[i][j], pomme_nb3[j][i] = pomme_nb3[j][i], pomme_nb3[i][j]
# vérification
affiche_image_nb(pomme_nb3)

**Q4** On cherche à obtenir le codage de l'image correspondant à une rotation d'un quart de tour (dans le sens trigonométrique) de l'image originale de la pomme.  

**On remarquera d'abord que** dans l'image obtenue par cette rotation les nombres de lignes et de colonnes sont échangés par rapport à la matrice de départ.

On construira la matrice codant l'image symétrique, nommée `pomme_nb4` à l'aide de <u>compréhensions de listes</u>, ***après avoir noté que les colonnes de la matrice de départ correspondent à des lignes de la matrice après rotation***.

Utiliser la fonction `affiche_image_nb` pour afficher l'image encodée par le tableau `pomme_nb4`.

***Q4 corrigé***

On doit d'abord préciser la remarque donnée comme indication, selon laquelle **les colonnes de la matrice de départ correspondent à des lignes de la matrice après rotation**.  

En effet, la colonne d'indice $0$ devient la ligne d'indice $p-1$, la colonne d'indice $1$ devient la ligne d'indice $p-2$, ..., la colonne d'indice $p-2$ devient la ligne d'indice $1$, la colonne d'indice $p-1$ devient la ligne d'indice $0$.

<img src ="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_matrice_avant_rotation.png" width ="800px"/>
<div style="text-align:center">Matrice avant rotation</div>

<img src ="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_matrice_apres_rotation.png" width ="600px"/>
<div style="text-align:center">Matrice après rotation</div>
<div style="text-align:center">Une colonne d'indice $j$ devient une ligne d'indice $j$</div>

De façon générale, la colonne d'indice $j$ devient la ligne d'indice $p-1-j$, ou, inversement, une ligne d'indice $i$ dans la matrice après rotation correspond à la colonne d'indice $p-1-i$ dans la matrice de départ.  
Plus précisément, si on note $m_{i,j}$ le pixel en ligne $i$ et colonne $j$ de la matrice avant rotation et $m'_{i',j'}$ le pixel en ligne $i'$ et colonne $j'$ de la matrice après rotation, la ligne d'indice $i'$, $0\leqslant i'<p$, dans $m'$ est la ligne composée des pixels suivants :  
$$
[m'_{i',0  }, m'_{i',1  }, \ldots, m'_{0,n-1}] = [m_{1,p-1-i'}, m_{i',p-1-i'}, \ldots, m_{n-1,p-1-i'}].
$$

À l'aide d'une compréhension de liste, la ligne `pomme_nb4[i']` s'écrirait donc `[pomme_nb[j][p-1-i'] for j in range(n)]`.

On en déduit la construction de la matrice `pomme_nb4` par compréhension de listes :

In [179]:
n, p = len(pomme_nb), len(pomme_nb[0]) ## <-- nombre de lignes, nombre de colonnes de la matrice pomme_nb
pomme_nb4 = [ [pomme_nb[j][p-1-i] for j in range(n)] for i in range(p)]

In [180]:
## vérification
affiche_image_nb(pomme_nb4)

On doit d'abord préciser la remarque donnée comme indication.  
Si on note $n$ et $p$, respectivement, les nombres de lignes et de colonne de l'image de départ, dans la rotation d'un quart de tour dans le sens trigonométrique, un pixel en ligne $i$, colonne $j$, devient le pixel en ligne $p - 1 - j$ et colonne $i$.  
Ainsi, si on note $m_{i,j}$ le pixel en ligne $i$ et colonne $j$ de la matrice avant rotation et $m'_{i',j'}$ le pixel en ligne $i'$ et colonne $j'$ de la matrice après rotation, la ligne d'indice $i'$, $0\leqslant i'<p$ est la ligne composée des pixels suivants :
$$
[m'_{i',0  }, m'_{i',0  }, \ldots, m'_{i',n-1}]
$$
on obtient la ligne d'indice $i$ de la matrice après rotation par la comp

In [181]:
affiche_image_nb(pomme_nb5)

Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'pomme_nb5' is not defined


**Q5** Reprendre le code de la question **Q4** pour définir une fonction, `rotation_quart`, prenant en argument un tableau, ***non nécessairement carré***, représentant une image en noir et blanc, et renvoyant un tableau codant l'image obtenue par une rotation d'un quart de tour (dans le sens trigonométrique) de l'image en entrée. 

Tester la fonction précédente sur un tableau précédemment construit et codant l'image d'une lettre.

Appliquer deux fois, puis quatre fois, la fonction `rotation_quart`, pour vérifier son bon fonctionnement.

***Q5 corrigé***

In [182]:
def rotation_quart(image):
    n, p = len(image), len(image[0]) ## <-- nombre de lignes, nombre de colonnes de la matrice image
    return [ [image[j][p-1-i] for j in range(n)] for i in range(p)]

* Test avec une lettre 

In [183]:
affiche_image_nb(rotation_quart(imageG_nb))

Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'imageG_nb' is not defined


Appliquer deux fois, puis quatre fois, la fonction `rotation_quart`, pour vérifier son bon fonctionnement.

In [184]:
affiche_image_nb(rotation_quart(rotation_quart(imageG_nb)))

Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'imageG_nb' is not defined


In [185]:
affiche_image_nb(rotation_quart(rotation_quart(rotation_quart(rotation_quart(imageG_nb)))))

Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'imageG_nb' is not defined


## Exercice 6 Manipulation d'une image en niveaux de gris

Pour définir une image en niveaux de gris, on peut utiliser pour les valeurs de chaque pixel, une échelle de gris, avec différentes intensités, allant du blanc au noir.  
En codant les niveaux de gris de 0 à 255 (le noir correspondant à la valeur 0 et le blanc à la valeur 255), on obtient 256 nuances de gris, dont les niveaux peuvent se coder sur 8 bits, ce qui est le compromis usuellement retenu pour avoir un nombre de nuances satisfaisant à l'œil sans encombrer la mémoire par le stockage de grands entiers.

Voici, un tableau codant suivant ce dernier principe une image de taille $8\times 8$, dont les 64 pixels sont à un niveau de gris aléatoire entre 0 et 255.

In [186]:
from random import randrange

aleaNG = [[randrange(2**8) for j in range(8)] for k in range(8)]

In [187]:
print(aleaNG)

[[12, 250, 166, 241, 206, 71, 61, 127], [222, 249, 102, 26, 121, 240, 209, 240], [26, 92, 83, 14, 140, 109, 169, 26], [179, 68, 94, 194, 56, 219, 164, 164], [105, 172, 176, 170, 30, 168, 180, 43], [137, 13, 16, 95, 78, 117, 252, 249], [154, 114, 1, 207, 161, 253, 90, 27], [28, 77, 3, 120, 159, 5, 18, 227]]


et que l'on peut encore afficher avec la fonction `affiche_image_nb`

In [188]:
affiche_image_nb(aleaNG)

**Q1** Reprendre le tableau, `pomme_nb`, codant l'image de pomme précédente, et en faire, à l'aide de compréhensions de listes, une copie profonde sous le nom `pomme_ng1`, en laissant les pixels noirs à zéro et en passant les pixels blancs à la valeur 255.

***Q1 corrigé***

In [189]:
pomme_ng1 = [[255 * pomme_nb[i][j] for j in range(len(pomme_nb[0]))] for i in range(len(pomme_nb))]

In [190]:
## vérification
affiche_image_nb(pomme_ng1)

**Q2** Modifier le tableau `pomme_ng1`, de sorte à ce que, pour tout $j$, $0\leqslant j < \text{len(pomme_ng[0])}$, les pixels noirs de la colonne d'indice $j$ aient pour intensité $j * 2^4$, les pixels blancs restant à la valeur $255$.

On devra obtenir le résultat suivant :<br/>
    <img src="https://github.com/lbaille20/Info1A_2021-22/raw/main/TP06/Capture_pomme_NG.png">

* Utiliser la fonction `affiche_image_nb` pour afficher l'image encodée par le tableau `pomme_ng1`.

***Q2 corrigé***

In [191]:
pomme_ng1 = [[255 * pomme_nb[i][j] for j in range(len(pomme_nb[0]))] for i in range(len(pomme_nb))]

for i in range(len(pomme_ng1)):
    for j in range(len(pomme_ng1[0])):
        if pomme_ng1[i][j] != 255:
            pomme_ng1[i][j] = j*2**4

In [192]:
affiche_image_nb(pomme_ng1)

## Exercice 7 Manipulation d'une image en couleurs

Pour définir une image en couleur, on peut utiliser pour la couleur de chaque pixel, un triplet de valeurs, comprises chacun entre 0 et 255, définissant, respectivement, l'intensité de rouge, de vert et de bleu, qui seront émises par les leds correspondant à chaque pixel.  

Voici, un tableau codant suivant ce dernier principe une image de taille $2\times 3$, dont les trois pixels de la première ligne, auront pour couleur les trois couleurs primaires, respectivement, rouge, vert et bleu, et dont les trois pixels de la seconde ligne, auront pour couleur les trois couleurs secondaires, respectivement, cyan, magenta et jaune, complémentaires des trois couleurs primaires :

In [193]:
imageRVB_base = [ [ [255,   0,   0], [  0, 255,   0], [  0,   0, 255] ],
                  [ [  0, 255, 255], [255,   0, 255], [255, 255,   0] ] ]

Un tableau codant suivant ce principe une image en couleur, sera ainsi un tableau à trois dimensions, de taille $n\times p \times c$, où $n$ est le nombre de lignes de l'image, $p$, son nombre de colonnes, et $c$ sont nombre de couleurs ($c=3$ pour une image codé en *RVB*).

On pourra visualiser une image en couleur à l'aide de la fonction suivante :

In [194]:
def affiche_image_rvb(tab):
    import matplotlib.pyplot as plt
    fig = plt.figure()
    plt.imshow(tab)
    plt.show()

On peut visualiser, par exemple, l'image codée par le tableau `imageRVB_base` :

In [195]:
affiche_image_rvb(imageRVB_base)

**Q1** Créer une image en couleurs $2\times 3$ du drapeau de la France, sous la forme d'un tableau, `drapeauFrance2x3`, de dimensions $2\times 3\times 3$, sachant que les couleurs du drapeau de la France sont codées en *RVB* par les niveaux de couleurs donnés ci-après :

In [196]:
bleu = [0, 35, 149]
blanc = [255, 255, 255]
rouge = [237, 41, 57]

* Faire appel à la fonction `affiche_image_rvb` pour vérifier que l'image représentée par le tableau `drapeauFrance2x3` est conforme aux attendus :

***Q1 corrigé***

In [197]:
drapeauFrance2x3 = [ [bleu, blanc, rouge] ] * 2

In [198]:
# vérification
drapeauFrance2x3

[[[0, 35, 149], [255, 255, 255], [237, 41, 57]], [[0, 35, 149], [255, 255, 255], [237, 41, 57]]]

In [199]:
affiche_image_rvb(drapeauFrance2x3)

**Q2** Créer une image du drapeau de la France de 200 sur 300 pixels sous la forme d'un tableau `drapeauFrance200x300`.  
***On pourra commencer par créer un tableau de zéros (entiers) de dimensions $200\times 300\times 3$, puis parcourir ce tableau suivant les lignes et les colonnes, pour changer les couleurs de pixels.***

* Faire appel à la fonction `affiche_image_rvb` pour vérifier que l'image représentée par le tableau `drapeauFrance200x300` est conforme aux attendus :

***Q2 corrigé***

In [200]:
n, p = 200, 300 # <-- nombres de lignes et de colonnes de l'image à construire
drapeauFrance200x300 = [[[0, 0, 0] for j in range(p)] for i in range(n)]
## coloriage de chacune des 200 lignes
for i in range(n):
    for j in range(p):
        if j < 100:
            drapeauFrance200x300[i][j] = bleu
        elif j < 200:
            drapeauFrance200x300[i][j] = blanc
        else:
            drapeauFrance200x300[i][j] = rouge
# ou encore :
drapeauFrance200x300 = [[bleu]*100 + [blanc]*100 + [rouge]*100]*200

In [201]:
# vérification
affiche_image_rvb(drapeauFrance200x300)

***Afin d'éviter de faire des tests sur la valeur de l'indice de colonne avant de colorier chaque pixel, on pourra préférer coloriser le drapeau colonne par colonne***, comme ci-dessous :

In [202]:
n, p = 200, 300 # <-- nombres de lignes et de colonnes de l'image à construire
drapeauFrance200x300 = [[[0, 0, 0] for j in range(p)] for i in range(n)]
## coloriage de chacune des 300 colonnes
for j in range(0, p // 3):
    for i in range(n):
        drapeauFrance200x300[i][j] = bleu
for j in range(p // 3, 2 * p // 3):
    for i in range(n):
        drapeauFrance200x300[i][j] = blanc
for j in range(2 * p // 3, p):
    for i in range(n):
        drapeauFrance200x300[i][j] = rouge

# vérification
affiche_image_rvb(drapeauFrance200x300)

Ou encore, en utilisant une liste des couleurs :

In [203]:
couleurs = [bleu, blanc, rouge]

n, p = 200, 300 # <-- nombres de lignes et de colonnes de l'image à construire
drapeauFrance200x300 = [[[0, 0, 0] for j in range(p)] for i in range(n)]
q = p // 3
for i in range(n):
    for j in range(p):
        drapeauFrance200x300[i][j] = couleurs[j // q]
# vérification
affiche_image_rvb(drapeauFrance200x300)

**[Préliminaire à la question Q3]**

On a téléchargé, depuis la page https://fr.wikipedia.org/wiki/Galerie_des_drapeaux_des_pays_du_monde, le drapeau de la Suède, dans un fichier `Flag_of_Sweden320x200.png`.

On pourra ouvrir ce fichier et récupérer l'image sous la forme d'un tableau en exécutant les deux cellules suivantes :

In [204]:
def charger_image(nomfichier_image):
    from PIL import Image
    im = Image.open(nomfichier_image)
    import numpy as np
    tab = np.asarray(im.convert("RGB"))
    return tab.tolist()

In [205]:
# ouverture de l'image sous la forme d'un tableau :
drapeauSuede = charger_image('Flag_of_Sweden320x200.png')

In [206]:
#examen du contenu du tableau
drapeauSuede

[[[0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106, 167], [0, 106,

* On affiche ci-dessous le drapeau de la Suède à l'aide de la fonction `affiche_image_rvb`

In [207]:
affiche_image_rvb(drapeauSuede)

**Q3** Caractéristiques du tableau `drapeauSuede`

**Q3.a** Écrire deux instructions permettant de vérifier que l'image se compose bien de 200 lignes de 320 pixels de couleur.  
On notera $n$ et $p$, respectivement, le nombre de lignes et de colonnes du tableau.

***Q3.a corrigé***

In [208]:
n = len(drapeauSuede) #nombre de lignes

In [209]:
p = len(drapeauSuede[0]) #nombre de pixels par ligne (nombre de colonnes du tableau)

In [210]:
n, p

(200, 320)

**Q3.b** Choisir deux pixels de l'image afin d'obtenir les codages RVB des deux couleurs du drapeau de la Suède.  
Vérifier à l'aide d'un parcours du tableau qu'il n'y a bien que deux couleurs différentes pour tous les pixels de l'image (on pourra utiliser l'opérateur `in`, ou `not in`, permettant de vérifier si une valeur est élément, ou non, d'une liste, d'un tuple ou d'un dictionnaire.  
S'il s'avère qu'il y a plus de deux couleurs dans le tableau, on construira un dictionnaire, `Dautres` dont les clés sont les tuples, $(r, v, b)$, codant les couleurs différentes des bleu et jaune majoritaires, associées à la liste des coordonnées des pixels de couleur $(r,v,b)$.  
Pour visualiser les pixels dont la couleur est différente des bleu et jaune majoritaires, on constuira une copie, `tab_autres` du tableau `drapeauSuede` dans laquelle tous les pixels seront mis à la couleur noire, sauf ceux dont la couleur est différente des bleu et jaune majoritaires.  
On affichera l'image correspondant à ce tableau.

***Q3.b corrigé***

In [211]:
#pour le bleu
drapeauSuede[0][0] ## couleur du pixel en haut à gauche

[0, 106, 167]

In [212]:
#pour le jaune
drapeauSuede[100][0] ## pixel située à l'extrême gauche sur une ligne "médiane", la ligne d'indice 100

[254, 204, 0]

On vérifie qu'il n'y a bien que deux couleurs dans le tableau, ici en construisant une liste, `Lautres`, des coordonnées, (ligne, colonne), des pixels qui ne seraient ni de couleur `bleu`, ni de couleur `jaune`.

In [213]:
bleu, jaune = drapeauSuede[0][0], drapeauSuede[100][0]

Lautres = []
for i in range(n):
    for j in range(p):
        if drapeauSuede[i][j] not in [bleu, jaune]:
            Lautres.append((i, j))

On constate que la liste `Lautres` n'est pas vide à l'issue du parcours, ce qui établit qu'il n'y a plus que deux couleurs dans l'image :

In [214]:
len(Lautres)

1756

In [215]:
Lautres

[(0, 99), (0, 100), (0, 139), (0, 140), (1, 99), (1, 100), (1, 139), (1, 140), (2, 99), (2, 100), (2, 139), (2, 140), (3, 99), (3, 100), (3, 139), (3, 140), (4, 99), (4, 100), (4, 139), (4, 140), (5, 99), (5, 100), (5, 139), (5, 140), (6, 99), (6, 100), (6, 139), (6, 140), (7, 99), (7, 100), (7, 139), (7, 140), (8, 99), (8, 100), (8, 139), (8, 140), (9, 99), (9, 100), (9, 139), (9, 140), (10, 99), (10, 100), (10, 139), (10, 140), (11, 99), (11, 100), (11, 139), (11, 140), (12, 99), (12, 100), (12, 139), (12, 140), (13, 99), (13, 100), (13, 139), (13, 140), (14, 99), (14, 100), (14, 139), (14, 140), (15, 99), (15, 100), (15, 139), (15, 140), (16, 99), (16, 100), (16, 139), (16, 140), (17, 99), (17, 100), (17, 139), (17, 140), (18, 99), (18, 100), (18, 139), (18, 140), (19, 99), (19, 100), (19, 139), (19, 140), (20, 99), (20, 100), (20, 139), (20, 140), (21, 99), (21, 100), (21, 139), (21, 140), (22, 99), (22, 100), (22, 139), (22, 140), (23, 99), (23, 100), (23, 139), (23, 140), (24, 99

Construction du dictionnaire `Dautres` :

In [216]:
Dautres = {}
for i in range(n):
    for j in range(p):
        if drapeauSuede[i][j] not in [bleu, jaune]:
            r, v, b = drapeauSuede[i][j]
            if (r, v, b) not in Dautres.keys():
                Dautres[(r, v, b)] = []
            Dautres[(r, v, b)].append((i, j))

Les couleurs différentes des bleu et jaune majoritaires sont données par :

In [217]:
Dautres.keys()

dict_keys([(0, 93, 187), (255, 216, 0), (0, 81, 208)])

In [218]:
bleu, jaune

([0, 106, 167], [254, 204, 0])

Construction du tableau `tab_autres`

In [219]:
bleu, jaune = drapeauSuede[0][0], drapeauSuede[100][0]

tab_autres = [ [ [0, 0, 0] for j in range(p) ] for i in range(n) ] 
for i in range(n):
    for j in range(p):
        if drapeauSuede[i][j] not in [bleu, jaune]:
            tab_autres[i][j] = drapeauSuede[i][j]

Affichage de l'image obtenue

In [220]:
affiche_image_rvb(tab_autres)

**Q3.c** Afin de corriger les couleurs de l'image d'origine, construire une copie, `drapeauSuede_corr`, du tableau `drapeauSuede`, dans laquelle, les couleurs différentes seront remplacées par les plus proches au sens de la distance euclidienne dans ${\mathbb R}^3$.  
On définira et on utilisera une fonction `distance(c1, c2)` renvoyant la distance distance euclidienne entre deux triplets `c1` et `c2`.

***Q3.c corrigé***

In [221]:
def distance3(c1, c2):
    x1, y1, z1 = c1
    x2, y2, z2 = c2
    return ((x2 - x1) ** 2 + (y2 - y1) ** 2 + (z2 - z1) ** 2) ** 0.5

bleu, jaune = drapeauSuede[0][0], drapeauSuede[100][0]

drapeauSuede_corr = [ [ drapeauSuede[i][j] for j in range(p) ] for i in range(n) ] 
for i in range(n):
    for j in range(p):
        couleur = drapeauSuede[i][j]
        if couleur not in [bleu, jaune]:
            if distance3(couleur, bleu) < distance3(couleur, jaune):
                drapeauSuede_corr[i][j] = bleu
            else:
                drapeauSuede_corr[i][j] = jaune

Affichage de l'image obtenue

In [222]:
affiche_image_rvb(drapeauSuede_corr)

**Q4** On cherche ici à reconstruire le tableau `drapeauSuede_corr`.

**Q4.a**  On cherche à déterminer les indices des lignes et des colonnes de couleur jaune grâce à un parcours du tableau `drapeauSuede_corr`.  
On construira pour cela un dictionnaire, `Dlignes`, dont les clés sont les indices de lignes, $i$, chaque clé $i$ étant associée à la liste des indices de colonnes, $j$, des pixels $(i, j)$ dont la couleur est différente de celle du pixel, $(i, j -1)$, qui les précède dans la ligne $i$.  
On cherchera ensuite, à l'aide d'un parcours par les valeurs du dictionnaire `Dlignes`, la liste `Lvarlignes` de toutes les liste d'indices de colonne différentes obtenues, puis, à l'aide d'un parcours par les couples (clé, valeur) du dictionnaire `Dlignes`, la liste `lignes_uniformes` de tous les indices de lignes sur lesquelles il n'y a aucun changement de couleur.  
Conclure en donnant tous les indices des lignes de couleur jaune et des colonnes de couleur jaune.

***Q4.a corrigé***

In [223]:
Dlignes = {}
for i in range(n):
    Dlignes[i] = []
    for j in range(1, p):
        if drapeauSuede_corr[i][j - 1] != drapeauSuede_corr[i][j]:
            Dlignes[i].append(j)

In [224]:
Dlignes

{0: [100, 140], 1: [100, 140], 2: [100, 140], 3: [100, 140], 4: [100, 140], 5: [100, 140], 6: [100, 140], 7: [100, 140], 8: [100, 140], 9: [100, 140], 10: [100, 140], 11: [100, 140], 12: [100, 140], 13: [100, 140], 14: [100, 140], 15: [100, 140], 16: [100, 140], 17: [100, 140], 18: [100, 140], 19: [100, 140], 20: [100, 140], 21: [100, 140], 22: [100, 140], 23: [100, 140], 24: [100, 140], 25: [100, 140], 26: [100, 140], 27: [100, 140], 28: [100, 140], 29: [100, 140], 30: [100, 140], 31: [100, 140], 32: [100, 140], 33: [100, 140], 34: [100, 140], 35: [100, 140], 36: [100, 140], 37: [100, 140], 38: [100, 140], 39: [100, 140], 40: [100, 140], 41: [100, 140], 42: [100, 140], 43: [100, 140], 44: [100, 140], 45: [100, 140], 46: [100, 140], 47: [100, 140], 48: [100, 140], 49: [100, 140], 50: [100, 140], 51: [100, 140], 52: [100, 140], 53: [100, 140], 54: [100, 140], 55: [100, 140], 56: [100, 140], 57: [100, 140], 58: [100, 140], 59: [100, 140], 60: [100, 140], 61: [100, 140], 62: [100, 140], 6

* recherche des listes différentes obtenues :

On construit la liste, `Lvarlignes` des listes distinctes obtenues :

In [225]:
Lvarlignes = []
for liste in Dlignes.values():
    if liste not in Lvarlignes:
        Lvarlignes.append(liste)

On obtient seulement deux listes distinctes :

In [226]:
Lvarlignes

[[100, 140], []]

* recherche des indices des lignes sans changement de couleur :

In [227]:
lignes_uniformes = []
for i, liste in Dlignes.items():
    if liste == []:
        lignes_uniformes.append(i)

La liste obtenue est la suivante :

In [228]:
lignes_uniformes

[80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119]

<i><b><u>En conclusion</u> :</b></i> sachant que le pixel en haut à gauche de l'image est bleu, on tire de l'examen de la liste `Lvarlignes` que les colonnes d'indices $j$ tel que $100\leqslant j< 140$, sont les colonnes de couleur jaune, et on tire de l'examen de la liste `lignes_uniformes`, que les lignes de couleur jaune sont les lignes d'indice $i$ tel que $80\leqslant i< 120$.

**Q4.b** Construire un tableau de pixels `tabSuede`, identique au tableau `drapeauSuede_corr` à l'aide des résultats obtenus à la question **Q4.a**.

* Faire appel à la fonction `affiche_image_rvb` pour vérifier que l'image représentée par le tableau `tabSuede` est conforme aux attendus :

***Q4.b corrigé***

Une solution est la suivante :

In [229]:
## hauteur et largeur de l'image :
n, p = len(drapeauSuede), len(drapeauSuede[0])
## couleurs dans l'image
bleu, jaune = drapeauSuede[0][0], drapeauSuede[100][0]
## les lignes uniformément jaunes sont les lignes d'indice i tel que i1 <= i < i2 où :
i1, i2 = lignes_uniformes[0], lignes_uniformes[0] + len(lignes_uniformes)
## les colonnes uniformément jaunes sont les colonnes d'indice j tel que j1 <= j < j2 où :
j1, j2 = Lvarlignes[0]

tabSuede = [[bleu for j in range(320)] for i in range(200)] ## construction du tableau tout en bleu

## coloriage de la bande horizontale jaune
for i in range(i1, i2):
    for j in range(p):
        tabSuede[i][j] = jaune
        
## coloriage de la bande verticale jaune
for j in range(j1, j2):
    for i in range(n):
        tabSuede[i][j] = jaune

On peut vérifier le résultat en affichant l'image :

In [230]:
affiche_image_rvb(tabSuede)

mais la vérification ultime consiste à comparer toutes les valeurs dans les deux tableaux :

In [231]:
tabSuede == drapeauSuede_corr

True